File: Source/mec/Source/Library/mec/src/main/java/com/philips/cdp/di/mec/integration/MECHandler.java
/**
 * (C) Koninklijke Philips N.V., 2015.
 * All rights reserved.
 */
package com.philips.cdp.di.mec.integration;

import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.FragmentTransaction;

import com.philips.cdp.di.ecs.ECSServices;
import com.philips.cdp.di.mec.common.MECFragmentLauncher;
import com.philips.cdp.di.mec.common.MECLauncherActivity;
import com.philips.cdp.di.mec.utils.MECConstant;
import com.philips.cdp.di.mec.utils.MECDataHolder;
import com.philips.cdp.di.mec.utils.MECutility;
import com.philips.platform.appinfra.AppInfra;
import com.philips.platform.appinfra.appconfiguration.AppConfigurationInterface;
import com.philips.platform.appinfra.servicediscovery.ServiceDiscoveryInterface;
import com.philips.platform.appinfra.servicediscovery.model.ServiceDiscoveryService;
import com.philips.platform.uappframework.launcher.ActivityLauncher;
import com.philips.platform.uappframework.launcher.FragmentLauncher;
import com.philips.platform.uappframework.launcher.UiLauncher;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.Objects;

class MECHandler {
    private MECDependencies mMECDependencies;
    private MECSettings mMECSetting;
    private UiLauncher mUiLauncher;
    private MECLaunchInput mLaunchInput;
    private AppInfra appInfra;
    private ArrayList<String> listOfServiceId;
    ServiceDiscoveryInterface.OnGetServiceUrlMapListener serviceUrlMapListener;
    private static final String IAP_PRIVACY_URL = "iap.privacyPolicy";
    private static final String IAP_FAQ_URL = "iap.faq";
    private static final String IAP_TERMS_URL = "iap.termOfUse";
    private static final String IAP_BASE_URL = "iap.baseurl";

    MECHandler(MECDependencies pMECDependencies, MECSettings pMecSettings, UiLauncher pUiLauncher, MECLaunchInput pLaunchInput) {
        this.mMECDependencies = pMECDependencies;
        this.mMECSetting = pMecSettings;
        this.mUiLauncher = pUiLauncher;
        this.mLaunchInput = pLaunchInput;

    }


    void launchMEC() {
        appInfra= (AppInfra) mMECDependencies.getAppInfra() ;
        AppConfigurationInterface configInterface = appInfra.getConfigInterface();
        AppConfigurationInterface.AppConfigurationError configError = new AppConfigurationInterface.AppConfigurationError();
        String propertyForKey = (String) configInterface.getPropertyForKey("propositionid", "MEC", configError);
        ECSServices ecsServices = new ECSServices(propertyForKey,appInfra);
        MecHolder.INSTANCE.eCSServices=ecsServices; // singleton
        MECDataHolder.INSTANCE.appinfra = appInfra;
        MECDataHolder.INSTANCE.propositionId = propertyForKey;
        MECDataHolder.INSTANCE.mecBannerEnabler = mLaunchInput.getMecBannerEnabler();
        MECDataHolder.INSTANCE.setHybrisEnabled(mLaunchInput.getHybrisEnabled());
        MECDataHolder.INSTANCE.mecBazaarVoiceInput = mLaunchInput.getMecBazaarVoiceInput();
        MECDataHolder.INSTANCE.blackListedRetailers = Objects.requireNonNull(mLaunchInput.getIgnoreRetailers());


        getUrl();
        if (mUiLauncher instanceof ActivityLauncher) {
            launchMECasActivity();
        } else {
            launchMECasFragment();
        }
    }

    void getUrl()
    {

        listOfServiceId = new ArrayList<>();
        listOfServiceId.add(IAP_BASE_URL);
        listOfServiceId.add(IAP_PRIVACY_URL);
        listOfServiceId.add(IAP_FAQ_URL);
        listOfServiceId.add(IAP_TERMS_URL);
        serviceUrlMapListener = new ServiceDiscoveryInterface.OnGetServiceUrlMapListener() {
            @Override
            public void onSuccess(Map<String, ServiceDiscoveryService> map) {
                Collection<ServiceDiscoveryService> collection = map.values();


                List<ServiceDiscoveryService> list = new ArrayList<>();
                list.addAll(collection);

                ServiceDiscoveryService discoveryService = map.get(IAP_PRIVACY_URL);
                assert discoveryService != null;
                String privacyUrl = discoveryService.getConfigUrls();
                if(privacyUrl != null) {
                    MECDataHolder.INSTANCE.setPrivacyUrl(privacyUrl);
                }

            }

            @Override
            public void onError(ERRORVALUES errorvalues, String s) {
                if (errorvalues.name().equals(ERRORVALUES.NO_NETWORK)) {
                }
            }
        };
        appInfra.getServiceDiscovery().getServicesWithCountryPreference(listOfServiceId, serviceUrlMapListener,null);
    }



    protected void launchMECasActivity() {
        Intent intent = new Intent(mMECSetting.getContext(), MECLauncherActivity.class);
        intent.putExtra(MECConstant.INSTANCE.getMEC_LANDING_SCREEN(), mLaunchInput.getMLandingView());
        ActivityLauncher activityLauncher = (ActivityLauncher) mUiLauncher;
        Bundle bundle =  getBundle();
        bundle.putInt(MECConstant.INSTANCE.getMEC_KEY_ACTIVITY_THEME(), activityLauncher.getUiKitTheme());
        intent.putExtras(bundle);
        mMECSetting.getContext().startActivity(intent);

    }

    protected void launchMECasFragment() {
        FragmentLauncher fragmentLauncher = (FragmentLauncher) mUiLauncher;
        Bundle bundle =  getBundle();
        bundle.putInt(MECConstant.INSTANCE.getMEC_LANDING_SCREEN(), mLaunchInput.getMLandingView());
        bundle.putInt("fragment_container", fragmentLauncher.getParentContainerResourceID()); // frame_layout for fragment
        loadDecisionFragment(bundle);


    }

     void loadDecisionFragment(Bundle bundle ){
        MECFragmentLauncher mecFragmentLauncher = new MECFragmentLauncher();
         mecFragmentLauncher.setArguments(bundle);

         FragmentLauncher fragmentLauncher = (FragmentLauncher)mUiLauncher;
         MECDataHolder.INSTANCE.setActionBarListener(fragmentLauncher.getActionbarListener(), mLaunchInput.getMecListener());
         String tag = mecFragmentLauncher.getClass().getName();
         FragmentTransaction transaction = fragmentLauncher.getFragmentActivity().getSupportFragmentManager().beginTransaction();
         transaction.replace(fragmentLauncher.getParentContainerResourceID(), mecFragmentLauncher, tag);
         transaction.commitAllowingStateLoss();

    }

    Bundle getBundle (){

        Bundle mBundle = new Bundle();
        if (mLaunchInput.getMMECFlowInput() != null) {
            if (mLaunchInput.getMMECFlowInput().getProductCTN() != null) {
                mBundle.putString(MECConstant.INSTANCE.getMEC_PRODUCT_CTN_NUMBER_FROM_VERTICAL(),
                        mLaunchInput.getMMECFlowInput().getProductCTN());
            }
            if (mLaunchInput.getMMECFlowInput().getProductCTNs() != null) {
                mBundle.putStringArrayList(MECConstant.INSTANCE.getCATEGORISED_PRODUCT_CTNS(),
                        mLaunchInput.getMMECFlowInput().getProductCTNs());
            }
            mBundle.putStringArrayList(MECConstant.INSTANCE.getMEC_IGNORE_RETAILER_LIST(), mLaunchInput.getIgnoreRetailers());
        }
        return mBundle;
    }


}

File: Source/mec/Source/Library/mec/src/main/java/com/philips/cdp/di/mec/integration/MECHandler.kt
/**
 * (C) Konin klijke Philips N.V., 2015.
 * All rights reserved.
 */
package com.philips.cdp.di.mec.integration

import android.app.Application
import android.content.Intent
import android.os.Bundle
import android.util.Log
import com.philips.cdp.di.ecs.error.ECSError
import com.philips.cdp.di.ecs.integration.ECSCallback
import com.philips.cdp.di.ecs.model.config.ECSConfig
import com.philips.cdp.di.mec.analytics.MECAnalytics
import com.philips.cdp.di.mec.common.MECLauncherActivity
import com.philips.cdp.di.mec.integration.serviceDiscovery.ServiceDiscoveryMapListener
import com.philips.cdp.di.mec.screens.reviews.BazaarVoiceHelper
import com.philips.cdp.di.mec.utils.MECConstant
import com.philips.cdp.di.mec.utils.MECDataHolder
import com.philips.platform.appinfra.AppInfra
import com.philips.platform.appinfra.servicediscovery.ServiceDiscoveryInterface.OnGetServiceUrlMapListener
import com.philips.platform.uappframework.launcher.ActivityLauncher
import com.philips.platform.uappframework.launcher.FragmentLauncher
import com.philips.platform.uappframework.launcher.UiLauncher
import java.util.*


internal open class MECHandler(private val mMECDependencies: MECDependencies, private val mMECSetting: MECSettings, private val mUiLauncher: UiLauncher, private val mLaunchInput: MECLaunchInput) {
    private var appInfra: AppInfra? = null
    private var listOfServiceId: ArrayList<String>? = null
    lateinit var serviceUrlMapListener: OnGetServiceUrlMapListener


    companion object {
        val IAP_PRIVACY_URL = "iap.privacyPolicy"
        val IAP_FAQ_URL = "iap.faq"
        val IAP_TERMS_URL = "iap.termOfUse"
    }

    // mBundle.putSerializable(MECConstant.FLOW_INPUT,mLaunchInput.getFlowConfigurator());
    fun getBundle(): Bundle {
        val mBundle = Bundle()
        if (mLaunchInput.flowConfigurator != null) {

            mBundle.putSerializable(MECConstant.FLOW_INPUT, mLaunchInput.flowConfigurator)

            if (mLaunchInput.flowConfigurator!!.productCTNs != null) {
                mBundle.putStringArrayList(MECConstant.CATEGORISED_PRODUCT_CTNS,
                        mLaunchInput.flowConfigurator!!.productCTNs)
            }
        }
        return mBundle
    }


    fun launchMEC() {
        /*  appInfra = mMECDependencies.appInfra as AppInfra
          val configInterface = appInfra!!.configInterface
          val configError = AppConfigurationInterface.AppConfigurationError()
          val propositionID = configInterface.getPropertyForKey("propositionid", "MEC", configError)
          var propertyForKey = ""
          if(propositionID!=null) {
              propertyForKey = propositionID as String
          }
          val ecsServices = ECSServices(propertyForKey, appInfra!!)
          MECDataHolder.INSTANCE.eCSServices = ecsServices // singleton
          MECDataHolder.INSTANCE.appinfra = appInfra as AppInfra
          MECDataHolder.INSTANCE.propositionId = propertyForKey*/
        MECDataHolder.INSTANCE.mecBannerEnabler = mLaunchInput.mecBannerConfigurator
        MECDataHolder.INSTANCE.hybrisEnabled = mLaunchInput.supportsHybris
        MECDataHolder.INSTANCE.retailerEnabled = mLaunchInput.supportsRetailer
        MECDataHolder.INSTANCE.mecBazaarVoiceInput = mLaunchInput.mecBazaarVoiceInput!!
        MECDataHolder.INSTANCE.voucherCode = mLaunchInput.voucherCode
        MECDataHolder.INSTANCE.maxCartCount = mLaunchInput.maxCartCount
        MECDataHolder.INSTANCE.mecBazaarVoiceInput = mLaunchInput.mecBazaarVoiceInput

        if (MECDataHolder.INSTANCE.bvClient == null) {
            val bazarvoiceSDK = BazaarVoiceHelper().getBazarvoiceClient(mMECSetting.context.applicationContext as Application)
            MECDataHolder.INSTANCE.bvClient = bazarvoiceSDK
        }

        MECDataHolder.INSTANCE.blackListedRetailers = mLaunchInput.blackListedRetailerNames

        MECAnalytics.initMECAnalytics(mMECDependencies)

        getUrl()

        // get config

        MECDataHolder.INSTANCE.eCSServices.configureECSToGetConfiguration(object : ECSCallback<ECSConfig, Exception> {

            override fun onResponse(config: ECSConfig?) {


                //set config data to singleton
                MECDataHolder.INSTANCE.config = config

                if (MECDataHolder.INSTANCE.hybrisEnabled) {
                    MECDataHolder.INSTANCE.hybrisEnabled = config?.isHybris ?: return
                }

                MECDataHolder.INSTANCE.locale = config!!.locale
                MECAnalytics.setCurrencyString(MECDataHolder.INSTANCE.locale)
                if (null != config!!.rootCategory) {
                    MECDataHolder.INSTANCE.rootCategory = config!!.rootCategory
                }

                // Launch fragment or activity
                if (mUiLauncher is ActivityLauncher) {
                    launchMECasActivity(MECDataHolder.INSTANCE.hybrisEnabled)
                } else {
                    launchMECasFragment(MECDataHolder.INSTANCE.hybrisEnabled)
                }
            }

            override fun onFailure(error: Exception?, ecsError: ECSError?) {
                Log.e("MEC", "Config failed : Can't Launch MEC")
            }
        })


    }

    //TODO
    private fun getUrl() {

        listOfServiceId = ArrayList()
        listOfServiceId!!.add(IAP_PRIVACY_URL)
        listOfServiceId!!.add(IAP_FAQ_URL)
        listOfServiceId!!.add(IAP_TERMS_URL)
        serviceUrlMapListener = ServiceDiscoveryMapListener()
        MECDataHolder.INSTANCE.appinfra.serviceDiscovery.getServicesWithCountryPreference(listOfServiceId, serviceUrlMapListener, null)
    }


    protected fun launchMECasActivity(isHybris: Boolean) {
        val intent = Intent(mMECSetting.context, MECLauncherActivity::class.java)
        val activityLauncher = mUiLauncher as ActivityLauncher
        val bundle = getBundle()
        bundle.putSerializable(MECConstant.KEY_FLOW_CONFIGURATION, mLaunchInput.flowConfigurator)
        bundle.putBoolean(MECConstant.KEY_IS_HYBRIS, isHybris)
        bundle.putInt(MECConstant.MEC_KEY_ACTIVITY_THEME, activityLauncher.uiKitTheme)
        intent.putExtras(bundle)
        mMECSetting.context.startActivity(intent)

    }

    private fun launchMECasFragment(hybris: Boolean) {

        val bundle = getBundle()

        val mecLandingFragment = FragmentSelector().getLandingFragment(hybris, mLaunchInput.flowConfigurator!!, bundle)
        val fragmentLauncher = mUiLauncher as FragmentLauncher
        bundle.putInt("fragment_container", fragmentLauncher.parentContainerResourceID) // frame_layout for fragment
        mecLandingFragment?.arguments = bundle


        MECDataHolder.INSTANCE.setActionBarListener(fragmentLauncher.actionbarListener, mLaunchInput.mecListener!!)
        val transaction = fragmentLauncher.fragmentActivity.supportFragmentManager.beginTransaction()
        transaction.replace(fragmentLauncher.parentContainerResourceID, mecLandingFragment!!, mecLandingFragment.getFragmentTag())
        transaction.addToBackStack(mecLandingFragment.getFragmentTag())
        transaction.commitAllowingStateLoss()
    }

}
